lib/rules/string.js

```diff
@@ -0,0 +1,407 @@
+        const options = params.args[0];
+        this._matches = [ipv4Regex, ipv6RegEx];
+        this._cidr = 'forbidden';
+        this._versions = ['ipv4', 'ipv6'];
+        if (!options) {
+            return;
+        }
+
+        if (options.versions) {
+            const vers = options.versions;
+            const matches = [];
+            if (vers.indexOf('ipv4') !== -1) {
+                matches.push(ipv4Regex);
+            }
+            if (vers.indexOf('ipv6') !== -1) {
+                matches.push(ipv6RegEx);
+            }
+            this._versions = vers;
+            this._matches = matches;
+        }
+        if (options.cidr) {
+            this._cidrAllowed = options.cidr === 'required' || options.cidr === 'optional';
+            this._cidrRequired = options.cidr === 'required';
+        }
+    }
+
+    validateSync(params) {
+        const res = this._matches.some(regEx => {
+            const match = params.value.match(regEx);
+            if (!match) {
+                return false;
+            }
+            if (match[1] && !this._cidrAllowed) {
+                return {
+                    reason: 'cidr-not-allowed',
+                };
+            }
+            if (!match[1] && this._cidrRequired) {
+                return {
+                    reason: 'cidr-required',
+                };
+            }
+            return true;
+        });
+        if (!res) {
+            return {
+                reason: 'not-matched',
+                allowed: this._versions,
+            };
+        }
+        return res;
+    }
+
+    static ruleName() {
+        return 'string.ip';
+    }
+}
+
+class URI extends SyncRule {
+
+    compile(params) {
+        const options = params.args[0];
+        if (!options) {
+            return;
+        }
+        this._schemes = options.schemes;
+        if (options.schemes) {
+            this._schemes = options.schemes.map(matcher => {
+                if (matcher instanceof RegExp) {
+                    return matcher;
+                }
+                if (typeof matcher === 'string') {
+                    return new RegExp(`^${escapeRegExp(matcher)}$`);
+                }
+                return matcher;
+            });
+        }
+        this._allowRelative = options.allowRelative;
+    }
+
+    validateSync(params) {
+        const res = url.parse(params.value);
+        if (
+            this._schemes &&
+            !this._schemes.some(matcher => matcher.test(res.protocol.slice(0, -1)))
+        ) {
+            return {
+                reason: 'scheme not allowed',
+                allowed: this._schemes,
+            };
+        }
+        if (!res.host) {
+            if (res.path) {
+                return this._allowRelative || {
+                    reason: 'relativ-not-allowed',
+                };
+            }
+            return false;
+        }
+        if (!isURL(params.value)) {
+            return false;
+        }
+        return true;
+    }
```

Comment:
The complexity in this function is fairly high, it can probably be broken up a bit